fix: Statistics summary timezone alignment for negative UTC users

Sjoerd-Bo3 opened the pull request:
## Summary
- Fixed timezone boundary alignment causing 0g carbs in summary while chart showed 75g
- Extended fix to all statistics types (Meal, TDD, Bolus) with same issue
- Ensures consistent day boundary calculations between chart display and summary data

## Description
This PR resolves issue #687 where users (particularly in negative UTC timezones like UTC-7) experienced discrepancies between statistics chart data and summary calculations. The main symptom was charts correctly showing 75g of carbs while the summary displayed 0g.

### Root Cause
The issue stemmed from inconsistent date boundary calculations:
- **Chart data**: Used complex timezone alignment logic in `StatChartUtils.visibleDateRange()`
- **Summary data**: Used simple `calendar.startOfDay()` grouping during data collection
- **Filter logic**: Used direct date comparison without day boundary alignment

This created misaligned date ranges that failed to match cached data, especially problematic for users in negative UTC offsets.

### Solution
Standardized all date boundary logic to use consistent `calendar.startOfDay()` approach:

1. **Simplified chart date range calculation** - Removed complex time-based alignment
2. **Fixed summary filtering** - Added day boundary alignment before filtering cached data  
3. **Applied to all statistics** - Extended fix to Meal, TDD, and Bolus statistics

## Test plan
- [x] Verify meal summary matches chart data for carbs
- [x] Test with negative UTC timezone scenarios (UTC-7)
- [x] Confirm TDD and Bolus summaries also display correctly
- [x] Check regional formatting changes don't affect results
- [x] Validate fix works across different time intervals (day/week/month)

Fixes #687

Sjoerd-Bo3 committed:
fix: Statistics meal summary timezone alignment issue

Fixed discrepancy between carb data displayed in main graph vs summary
by ensuring consistent day boundary calculations between chart data and
summary calculations.

Key changes:
- Simplified visibleDateRange() to use same startOfDay() logic as meal data grouping
- Removed complex time-based alignment that caused issues for negative UTC timezones
- Added day boundary alignment in calculateAveragesForDateRange()
- Ensures both chart display and summary use identical date boundaries

Fixes #687

Sjoerd-Bo3 committed:
fix: Apply timezone alignment fix to TDD and Bolus statistics summaries

Extended the timezone boundary fix to TDD and Bolus statistics which had
the same date alignment issue as meal statistics. All summary calculations
now use consistent startOfDay() logic matching the chart data grouping.

- Fixed calculateTDDAveragesForDateRange() in TDDSetup.swift
- Fixed calculateBolusAveragesForDateRange() and calculateBolusTotalsForDateRange() in BolusStatsSetup.swift
- Ensures all statistics summaries work correctly for users in negative UTC timezones

Related to #687

kingst reviewed (commented):
Tested it in my simulator and it appears to be working. I left one comment, but there are a few -1 for dates can you double check to make sure that the aligning logic isn't dropping a day at the end?

Sjoerd-Bo3 committed:
fix: Remove forced unwraps from TDD storage date calculations

- Replace sortedEvents.first\! with safe guard statement
- Use TimeInterval.hours(24) for clearer date arithmetic
- Addresses kingst's feedback about potential crashes and timezone handling

This ensures the app won't crash with empty event arrays and makes
the date calculation logic more explicit and safer.

Sjoerd-Bo3 committed:
test: Add timezone tests for TDD storage addressing kingst's concerns

- Test -1 second logic across full UTC offset spectrum (-11 to +14)
- Verify empty events are handled without forced unwrap crashes
- Test DST transitions maintain correct day boundaries
- Ensure multiple consecutive days work correctly

Tests specifically verify kingst's concern that the -1 second
adjustment doesn't accidentally drop days with negative UTC offsets.

Signed-off-by: Sjoerd Bozon <sjoerd.bozon@gmail.com>

Sjoerd-Bo3 commented on Trio/Sources/Modules/Stat/View/StatChartUtils.swift:
```diff
@@ -31,21 +31,19 @@ struct StatChartUtils {
+            let endOfDay = calendar.startOfDay(for: endDate)
+            
+            // Ensure we include the full end day by going to the end of that day
+            let alignedEnd = calendar.date(byAdding: .day, value: 1, to: endOfDay)!.addingTimeInterval(-1)
```
Sure Fixed!

Sjoerd-Bo3 commented:
@kingst, I've addressed your feedback about the forced unwraps and the concern about the -1 logic potentially dropping days. Thanks for catching that!

## Changes Made

  1. Removed forced unwraps: Replaced sortedEvents.first!.timestamp with a safe guard statement that returns an empty array if there are no events
  2. Used clearer date arithmetic: Changed to TimeInterval.hours(24) as you suggested

## Testing Your Concern

  I added comprehensive tests specifically for your worry about the -1 second adjustment. The test testMinusOneLogicAcrossAllTimezones() verifies that:

  - The -1 logic works correctly across all UTC offsets (from UTC-11 to UTC+14)
  - Tests specifically stress-test with late evening times (23:30) when timezone differences are most pronounced
  - Verifies that startOfDay and endOfDay always remain on the same calendar day
  - Confirms we get exactly 23:59:59 as the end time

  For example, in Hawaii (UTC-10), when it's 11:30 PM on Jan 15, it's already 9:30 AM Jan 16 in UTC. The tests confirm that our calculation still correctly produces Jan 15 00:00:00 to Jan 15 23:59:59, not accidentally dropping back to
  Jan 14.

  Also tested:
  - DST transitions (spring forward/fall back)
  - Half-hour offset timezones like India (UTC+5:30)
  - Multiple consecutive days

  The fix is split into two commits:
  1. The actual code fix removing forced unwraps
  2. The comprehensive test suite

  Let me know if you'd like me to adjust anything else!

kingst reviewed (commented):
Looks good! I'm still seeing a forced unwrap on line 44 of Trio/Sources/Modules/Stat/View/StatChartUtils.swift but once you fix that I can approve

Sjoerd-Bo3 committed:
fix: Remove forced unwrap from StatChartUtils date calculation

Address kingst's code review feedback by replacing forced unwrap with nil coalescing operator on line 44, providing a safe fallback to endOfDay if the date calculation fails.

kingst reviewed (approved):
LGTM, thanks for taking this on!

MikePlante1 reviewed (changes requested):
The ending of visible date is incorrect with this PR applied. It should say Jul 2 - Jul 8, but instead it says Jul 2 - Jul 9. I checked the Week charts for TDD, Meals, Bolus Distribution, Percentile (by day), and Distribution (by day), and they all showed Jul 2 - Jul 9.

| start | end |
| --- | --- |
| ![IMG_7789](https://github.com/user-attachments/assets/22f46841-5b9a-445a-8fe9-1ba719053bd8) |  ![IMG_7788](https://github.com/user-attachments/assets/b692cfa8-6848-4a30-ab8a-825233658194) |

mountrcg commented:
@Sjoerd-Bo3 can confirm @MikePlante1 comment. It is less critical than what you fixed, but most probably can be remedied by setting that end date -1?

dnzxy commented:
Status here?

dnzxy commented:
3 months later, status here?

mountrcg commented:
I have been using this for the 3 months, the issue with d+1 being shown in the the time intervall for week or month etc.. Mike mentions is still there. But it’s the lesser evil compared to what was being fixed.

dnzxy commented:
So this PR isn’t done cause we don’t fix one thing to break another. 

Please fix the date offset, then this can move up.

dnzxy commented:
Can we please move this forward one way or the other @Sjoerd-Bo3 ? If you don't have the resources to doing so, maybe @mountrcg can help with the last issue that was identified during testing.

dnzxy commented:
@Sjoerd-Bo3 @mountrcg Can you please coordinate on how to move forward on this PR? It's been sitting stale for about 5 months now. Please address the changes requested by @MikePlante1 .

dnzxy commented:
Closing this for now. @MikePlante1 will take a look when resources free up on his end.
